compiler/src/dotty/tools/dotc/transform/LazyVals.scala

```diff
@@ -274,168 +282,212 @@ class LazyVals extends MiniPhase with IdentityDenotTransformer {
+        initBlock(ValDef(lockSymb, ref(target).cast(waiting.typeOpt))
+          :: objCasFlag.appliedTo(thiz, offset, ref(lockSymb), ref(resSymb))
+          :: ref(lockSymb).select(lazyNme.RLazyVals.waitingRelease).ensureApplied :: Nil)
+      }
+      // finally block
+      val fin = If(
+        objCasFlag.appliedTo(thiz, offset, evaluating, ref(resSymb)).equal(Literal(Constant(false))), 
+        lockRel,
+        EmptyTree
+      ).withType(defn.UnitType)
+      // entire try block
+      val evaluate = Try(
+        initBlock(
+          Assign(ref(resSymb), if needsBoxing(tp) && rhs != EmptyTree then rhs.ensureConforms(boxIfCan(tp)) else rhs) // try result = rhs
+          :: nullOut(nullableFor(methodSymbol))
+          ::: If(ref(resSymb).equal(nullLiteral), Assign(ref(resSymb), nullValued), EmptyTree).withType(defn.UnitType) // if result == null then result = NULL
+          :: Nil
+        ),
+        Nil,
+        fin
       )
+      // if CAS(...)
+      If(
+        objCasFlag.appliedTo(thiz, offset, nullLiteral, evaluating),
+        initBlock(ValDef(resSymb, nullLiteral) // var result: AnyRef = null
+          :: evaluate // try ... finally ...
+          :: Nil),
+        EmptyTree
+      ).withType(defn.UnitType)
     }
-
-    val initialize = If(
-      casFlag.appliedTo(thiz, offset, flagRef, computeState, fieldId),
-      Try(compute, List(retryCase), EmptyTree),
-      unitLiteral
-    )
-
-    val condition = If(
-      stateRef.equal(computedState),
-      Return(ref(target), methodSymbol),
+    val current = newSymbol(methodSymbol, lazyNme.current, Synthetic, defn.ObjectType)
+    val ifNotNull =
       If(
-        stateRef.equal(initState),
-        initialize,
-        waitOnLock.appliedTo(thiz, offset, flagRef, fieldId)
+        ref(current).select(defn.Any_isInstanceOf).appliedToTypeTree(evaluating),
+        ref(discardSymb).ensureApplied,
+        // not an Evaluating
+        If(
+          ref(current).select(defn.Any_isInstanceOf).appliedToTypeTree(nullValued),
+          Return(defaultValue(tp), methodSymbol),
+          // not a NULL
+          If(
+            ref(current).select(defn.Any_isInstanceOf).appliedToTypeTree(waiting),
+            ref(current).select(defn.Any_asInstanceOf).appliedToTypeTree(waiting).select(lazyNme.RLazyVals.waitingAwaitRelease).ensureApplied,
+            // not a Waiting, then is an A
+            Return(ref(current).ensureConforms(tp), methodSymbol)
+          )
+        )
       )
-    )
-
-    val loop = WhileDo(EmptyTree, Block(List(flagDef, stateDef), condition))
-    DefDef(methodSymbol, loop)
+    val body = initBlock(ValDef(current, ref(target)) :: If(ref(current).equal(nullLiteral), unevaluated, ifNotNull) :: Nil)
+    val mainLoop = WhileDo(EmptyTree, body) // becomes: while (true) do { body }
+    val ret = DefDef(methodSymbol, initBlock(discardDef :: mainLoop :: Nil))
+    ret
   }
 
   def transformMemberDefThreadSafe(x: ValOrDefDef)(using Context): Thicket = {
```

Comment:
When targetting an older version such as 3.0 we need to generate the old version of the code.
I suggest making two variants of `transformMemberDefThreadSafe` one with the old code and one with the new code. Do the same for `mkThreadSafeDef`. Then `transformMemberDefThreadSafe` can dispatch to the correct one as follows
```scala
import dotty.tools.dotc.config.ScalaRelease.*
def transformMemberDefThreadSafe(x: ValOrDefDef)(using Context): Thicket =
  // TODO find more meaningful names, not old/new.
  if ctx.scalaRelease <= Release3_1 then transformMemberDefThreadSafeOld(x) 
  else transformMemberDefThreadSafeNew(x)
```

Note: use `Release3_0` as a starting point. Otherwise, you will always get the old version as the main is also on 3.1. This is something that we have to change when going to 3.2.

Reply:
You can use a variant of `lazyVals_c3.0.0.scala` named `lazyVals_r3.0.scala` to have a test that sets the scala release flag to 3.0. This will compile to code with the current compiler but target the old stdlib. Unfortunately, the test will run with the current library. We still need to figure out the simplest way to run with the old library.